src/aws_encryption_sdk/materials_managers/__init__.py

```diff
@@ -40,38 +48,254 @@ class EncryptionMaterialsRequest(object):
-    plaintext_length = attr.ib(
-        default=None, validator=attr.validators.optional(attr.validators.instance_of(six.integer_types))
+    data_encryption_key = attr.ib(
+        default=None, validator=optional(instance_of(RawDataKey)), converter=_data_key_to_raw_data_key
     )
+    _keyring_trace = attr.ib(
+        default=attr.Factory(list), validator=optional(deep_iterable(member_validator=instance_of(KeyringTrace)))
+    )
+    _initialized = False
 
+    def __attrs_post_init__(self):
+        """Freeze attributes after initialization."""
+        self._initialized = True
 
-@attr.s(hash=False)
-class EncryptionMaterials(object):
+    def __setattr__(self, key, value):
+        """Do not allow attributes to be changed once an instance is initialized."""
+        if self._initialized:
+            raise AttributeError("can't set attribute")
+
+        self._setattr(key, value)
+
+    def _setattr(self, key, value):
+        """Special __setattr__ to avoid having to perform multi-level super calls."""
+        super(CryptographicMaterials, self).__setattr__(key, value)
+
+    def _validate_data_encryption_key(self, data_encryption_key, keyring_trace, required_flags):
+        # type: (Union[DataKey, RawDataKey], KeyringTrace, Iterable[KeyringTraceFlag]) -> None
+        """Validate that the provided data encryption key and keyring trace match for each other and the materials.
+
+        :param RawDataKey data_encryption_key: Data encryption key
+        :param KeyringTrace keyring_trace: Keyring trace corresponding to data_encryption_key
+        :param required_flags: Iterable of required flags
+        :type required_flags: iterable of :class:`KeyringTraceFlag`
+        :raises AttributeError: if data encryption key is already set
+        :raises InvalidKeyringTraceError: if keyring trace does not match decrypt action
+        :raises InvalidKeyringTraceError: if keyring trace does not match data key provider
+        :raises InvalidDataKeyError: if data key length does not match algorithm suite
+        """
+        if self.data_encryption_key is not None:
+            raise AttributeError("Data encryption key is already set.")
+
+        for flag in required_flags:
+            if flag not in keyring_trace.flags:
+                raise InvalidKeyringTraceError("Keyring flags do not match action.")
+
+        if keyring_trace.wrapping_key != data_encryption_key.key_provider:
+            raise InvalidKeyringTraceError("Keyring trace does not match data key provider.")
+
+        if len(data_encryption_key.data_key) != self.algorithm.kdf_input_len:
+            raise InvalidDataKeyError(
+                "Invalid data key length {actual} must be {expected}.".format(
+                    actual=len(data_encryption_key.data_key), expected=self.algorithm.kdf_input_len
+                )
+            )
+
+    def _add_data_encryption_key(self, data_encryption_key, keyring_trace, required_flags):
```

Comment:
I keep reading this name and confusing it.

`data_encryption_key` and `encrypted_data_key` seem a little close?

Reply:
You're not wrong, but that's what the API has been since we introduced CMMs. Either we need to change the API with backwards compatible shims (see `DecryptionMaterials`) or we can just accept the similar names.

Reply:
sadness

Reply:
Unless this is causing customer confusion, I don't think so. This would be a non-trivial change for (IMO) not much benefit.